pkg/square/builder.go

```diff
@@ -0,0 +1,238 @@
+package square
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sort"
+
+	"github.com/celestiaorg/celestia-app/pkg/appconsts"
+	"github.com/celestiaorg/celestia-app/pkg/namespace"
+	"github.com/celestiaorg/celestia-app/pkg/shares"
+	"github.com/tendermint/tendermint/pkg/consts"
+	coretypes "github.com/tendermint/tendermint/proto/tendermint/types"
+	core "github.com/tendermint/tendermint/types"
+)
+
+type Builder struct {
+	// upper bound on the total amount of shares that could fit in the biggest square
+	maxCapacity int
+	// current estimated amount of shares used. These are worst case estimates.
+	currentSize int
+
+	// here we keep track of the actual data to go in a square
+	txs   [][]byte
+	pfbs  []*coretypes.IndexWrapper
+	blobs []*element
+
+	// for compact shares we use a counter to track the amount of shares needed
+	txCounter  *shares.CompactShareCounter
+	pfbCounter *shares.CompactShareCounter
+}
+
+func NewBuilder(maxSquareSize int) (*Builder, error) {
+	if maxSquareSize < 0 {
+		return nil, errors.New("max square size must be positive")
+	}
+	if !shares.IsPowerOfTwo(maxSquareSize) {
+		return nil, errors.New("max square size must be a power of two")
+	}
+	return &Builder{
+		maxCapacity: maxSquareSize * maxSquareSize,
+		blobs:       make([]*element, 0),
+		pfbs:        make([]*coretypes.IndexWrapper, 0),
+		txs:         make([][]byte, 0),
+		txCounter:   shares.NewCompactShareCounter(),
+		pfbCounter:  shares.NewCompactShareCounter(),
+	}, nil
+}
+
+// AppendTx attempts to allocate the transaction to the square. It returns false if there is not
+// enough space in the square to fit the transaction.
+func (c *Builder) AppendTx(tx []byte) bool {
+	lenChange := c.txCounter.Add(len(tx))
+	if c.canFit(lenChange) {
+		c.txs = append(c.txs, tx)
+		c.currentSize += lenChange
+		return true
+	}
+	c.txCounter.Revert()
+	return false
+}
+
+// AppendBlobTx attempts to allocate the blob transaction to the square. It returns false if there is not
+// enough space in the square to fit the transaction.
+func (c *Builder) AppendBlobTx(blobTx coretypes.BlobTx) bool {
+	iw := &coretypes.IndexWrapper{
+		Tx:           blobTx.Tx,
+		TypeId:       consts.ProtoIndexWrapperTypeID,
+		ShareIndexes: worstCaseShareIndexes(len(blobTx.Blobs), c.maxCapacity),
+	}
+	size := iw.Size()
+	pfbShareDiff := c.pfbCounter.Add(size)
```

Comment:
nice, this is a significant improvement to https://github.com/celestiaorg/celestia-app/blob/f31ee58890b1f94815057b4bbd77d7cbe62172f4/app/estimate_square_size.go#L74-L105